File: frontend/src-tauri/src/database/repositories/ask_conversation.rs
use sqlx::SqlitePool;

use crate::database::models::{AskConversationRow, AskMessageRow};

pub struct AskConversationRepository;

impl AskConversationRepository {
    /// Delete conversations (and their messages) not touched since `cutoff_rfc3339`.
    /// Enforces the 7-day retention window; called before list/append.
    pub async fn prune_older_than(
        pool: &SqlitePool,
        cutoff_rfc3339: &str,
    ) -> Result<(), sqlx::Error> {
        let mut tx = pool.begin().await?;
        sqlx::query(
            "DELETE FROM ask_messages WHERE conversation_id IN \
             (SELECT id FROM ask_conversations WHERE updated_at < ?)",
        )
        .bind(cutoff_rfc3339)
        .execute(&mut *tx)
        .await?;
        sqlx::query("DELETE FROM ask_conversations WHERE updated_at < ?")
            .bind(cutoff_rfc3339)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;
        Ok(())
    }

    pub async fn create(
        pool: &SqlitePool,
        id: &str,
        meeting_id: Option<&str>,
        title: Option<&str>,
        now_rfc3339: &str,
    ) -> Result<(), sqlx::Error> {
        sqlx::query(
            "INSERT INTO ask_conversations (id, meeting_id, title, created_at, updated_at) \
             VALUES (?, ?, ?, ?, ?)",
        )
        .bind(id)
        .bind(meeting_id)
        .bind(title)
        .bind(now_rfc3339)
        .bind(now_rfc3339)
        .execute(pool)
        .await?;
        Ok(())
    }

    /// List conversations for a scope, most-recently-updated first. `meeting_id = None`
    /// returns global (NULL-meeting) conversations; `Some(id)` returns that meeting's.
    pub async fn list(
        pool: &SqlitePool,
        meeting_id: Option<&str>,
    ) -> Result<Vec<AskConversationRow>, sqlx::Error> {
        match meeting_id {
            Some(meeting_id) => {
                sqlx::query_as::<_, AskConversationRow>(
                    "SELECT id, meeting_id, title, created_at, updated_at FROM ask_conversations \
                     WHERE meeting_id = ? ORDER BY updated_at DESC",
                )
                .bind(meeting_id)
                .fetch_all(pool)
                .await
            }
            None => {
                sqlx::query_as::<_, AskConversationRow>(
                    "SELECT id, meeting_id, title, created_at, updated_at FROM ask_conversations \
                     WHERE meeting_id IS NULL ORDER BY updated_at DESC",
                )
                .fetch_all(pool)
                .await
            }
        }
    }

    pub async fn get(
        pool: &SqlitePool,
        id: &str,
    ) -> Result<Option<AskConversationRow>, sqlx::Error> {
        sqlx::query_as::<_, AskConversationRow>(
            "SELECT id, meeting_id, title, created_at, updated_at FROM ask_conversations WHERE id = ?",
        )
        .bind(id)
        .fetch_optional(pool)
        .await
    }

    pub async fn messages(
        pool: &SqlitePool,
        conversation_id: &str,
    ) -> Result<Vec<AskMessageRow>, sqlx::Error> {
        sqlx::query_as::<_, AskMessageRow>(
            "SELECT id, conversation_id, role, content, sources_json, created_at \
             FROM ask_messages WHERE conversation_id = ? ORDER BY created_at ASC",
        )
        .bind(conversation_id)
        .fetch_all(pool)
        .await
    }

    /// Append a message and bump the conversation's `updated_at` (one transaction).
    pub async fn append_message(
        pool: &SqlitePool,
        message_id: &str,
        conversation_id: &str,
        role: &str,
        content: &str,
        sources_json: Option<&str>,
        now_rfc3339: &str,
    ) -> Result<(), sqlx::Error> {
        let mut tx = pool.begin().await?;
        sqlx::query(
            "INSERT INTO ask_messages (id, conversation_id, role, content, sources_json, created_at) \
             VALUES (?, ?, ?, ?, ?, ?)",
        )
        .bind(message_id)
        .bind(conversation_id)
        .bind(role)
        .bind(content)
        .bind(sources_json)
        .bind(now_rfc3339)
        .execute(&mut *tx)
        .await?;
        sqlx::query("UPDATE ask_conversations SET updated_at = ? WHERE id = ?")
            .bind(now_rfc3339)
            .bind(conversation_id)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;
        Ok(())
    }

    pub async fn set_title(
        pool: &SqlitePool,
        id: &str,
        title: &str,
        now_rfc3339: &str,
    ) -> Result<(), sqlx::Error> {
        sqlx::query("UPDATE ask_conversations SET title = ?, updated_at = ? WHERE id = ?")
            .bind(title)
            .bind(now_rfc3339)
            .bind(id)
            .execute(pool)
            .await?;
        Ok(())
    }

    pub async fn delete(pool: &SqlitePool, id: &str) -> Result<(), sqlx::Error> {
        let mut tx = pool.begin().await?;
        sqlx::query("DELETE FROM ask_messages WHERE conversation_id = ?")
            .bind(id)
            .execute(&mut *tx)
            .await?;
        sqlx::query("DELETE FROM ask_conversations WHERE id = ?")
            .bind(id)
            .execute(&mut *tx)
            .await?;
        tx.commit().await?;
        Ok(())
    }
}

File: AriKit/Sources/AriKit/Recall/Conversations/AskConversation.swift
//
//  AskConversation.swift — public domain values for the Ask conversation store (plan §2.3,
//  ← ari-engine/src/recall/conversations.rs `AskConversationDto`/`AskMessageDto`/
//  `AskConversationDetailDto`, database/models.rs `AskConversationRow`/`AskMessageRow`).
//
//  Zero DB dependency — plain `Sendable` value types; `AskConversationStore` translates them
//  to/from the internal GRDB records in this same directory. Timestamps stay raw RFC3339
//  `String` (not `Date`), matching the Recall Slice 2 index tables' convention
//  (`Recall/Index/RecallChunk.swift`) since the `askConversation`/`askMessage` columns are
//  GRDB `.text`, not `.datetime` (`SchemaMigrator.swift`).
//
import Foundation

public typealias AskConversationID = Identifier<AskConversation>
public typealias AskMessageID = Identifier<AskMessage>

/// A conversation's header row (← Rust `AskConversationDto`, conversations.rs:24-30, extended by
/// `docs/plans/ari-ask-ui.md` Phase 0). Scope invariant: **at most one of `meetingId`/`seriesId`
/// is non-nil**; both nil is a global (cross-meeting, cross-series) chat. `AskConversationStore`
/// enforces this on `create`/`list`, not a DB constraint.
public struct AskConversation: Codable, Hashable, Sendable, Identifiable {
    public var id: AskConversationID
    public var meetingId: MeetingID?
    public var seriesId: SeriesID?
    public var title: String?
    public var createdAt: String
    public var updatedAt: String

    public init(
        id: AskConversationID,
        meetingId: MeetingID? = nil,
        seriesId: SeriesID? = nil,
        title: String? = nil,
        createdAt: String,
        updatedAt: String
    ) {
        self.id = id
        self.meetingId = meetingId
        self.seriesId = seriesId
        self.title = title
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

/// One turn in a conversation (← Rust `AskMessageDto`, conversations.rs:46-54). `sources` is
/// APP-SUPPLIED — built by the orchestrator from the DB, never parsed back from model output
/// (plan §7) — and is `[]`, never `nil`, when there is nothing real to show (No-Fake-State):
/// unlike Rust's `Option<Vec<LocalRecallSource>>` (whose `None` and `Some([])` are
/// distinguishable), this port collapses "no sources" to an honest empty array on both write and
/// read — a documented, deliberate delta from the Rust wire shape.
public struct AskMessage: Codable, Hashable, Sendable, Identifiable {
    public var id: AskMessageID
    public var conversationId: AskConversationID
    public var role: String
    public var content: String
    public var sources: [RecallSource]
    /// A deterministically-resolved entity card (`ask-meetings-tools-and-cards.md` §5.1), additive.
    /// `nil` for every message except an assistant turn whose ask resolved exactly one real entity
    /// — never a partial match, never a placeholder (No-Fake-State). Kept as back-compat, always
    /// `cards.first` (plan §5.4, `ask-meetings-agentic-tools.md`).
    public var card: RecallCardPayload?
    /// The full set of resolved cards (plan §5.4) — the tool-first agentic path can resolve more
    /// than one entity per ask. `[]` for every message except an assistant turn that resolved at
    /// least one real entity.
    public var cards: [RecallCardPayload]
    public var createdAt: String

    public init(
        id: AskMessageID,
        conversationId: AskConversationID,
        role: String,
        content: String,
        sources: [RecallSource] = [],
        card: RecallCardPayload? = nil,
        cards: [RecallCardPayload]? = nil,
        createdAt: String
    ) {
        self.id = id
        self.conversationId = conversationId
        self.role = role
        self.content = content
        self.sources = sources
        let resolvedCards = cards ?? (card.map { [$0] } ?? [])
        self.cards = resolvedCards
        self.card = card ?? resolvedCards.first
        self.createdAt = createdAt
    }
}

/// A conversation with its full message history (← Rust `AskConversationDetailDto`,
/// conversations.rs:75-78), ordered oldest-first (matches the Rust `messages` query's
/// `ORDER BY created_at ASC`, `ask_conversation.rs:96`).
public struct AskConversationDetail: Codable, Hashable, Sendable {
    public var conversation: AskConversation
    public var messages: [AskMessage]

    public init(conversation: AskConversation, messages: [AskMessage]) {
        self.conversation = conversation
        self.messages = messages
    }
}
